File: inspect_app_yamazaki.js
const fs = require('fs');

// Mock window/localStorage
global.window = {};
global.localStorage = {
    getItem: () => null,
    setItem: () => null
};

// Read data.js
const dataJs = fs.readFileSync('data.js', 'utf8');
eval(dataJs);

const transfers = window.initialTransfers;
console.log(`Loaded ${transfers.length} transfers from data.js.`);

const appJs = fs.readFileSync('app.js', 'utf8');

// Extract linkTransfers function
const linkTransfersMatch = appJs.match(/function linkTransfers[\s\S]*?return \[\.\.\.originals, \.\.\.correctives\];\s*\n?\s*\}/);
if (!linkTransfersMatch) {
    console.log("Could not find linkTransfers in app.js");
    process.exit(1);
}
eval(linkTransfersMatch[0]);

// Extract calculateStatus function
const calculateStatusMatch = appJs.match(/function calculateStatus[\s\S]*?return \{\s*slChuyenKRC[\s\S]*?badgeClass\s*\}\s*;\s*\n?\s*\}/);
if (!calculateStatusMatch) {
    // try another pattern
    const match2 = appJs.match(/function calculateStatus[\s\S]*?statusText,[\s\S]*?badgeClass[\s\S]*?\}/);
    if (match2) {
        eval(match2[0]);
    } else {
        console.log("Could not find calculateStatus in app.js");
        process.exit(1);
    }
} else {
    eval(calculateStatusMatch[0]);
}

// Run linking
const processedTransfers = linkTransfers(transfers);
console.log(`Processed ${processedTransfers.length} transfers.`);

// Filter for Yamazaki items
const codes = ['SP001222', 'SP001988'];
const yamazakiTransfers = processedTransfers.filter(t => codes.includes(t.itemCode));

console.log("\n--- YAMAZAKI TRANSFERS AND CALCULATED STATUS ---");
yamazakiTransfers.forEach(t => {
    // Only print if original with difference, or corrective
    const status = calculateStatus(t);
    const fromB = t.fromBranch;
    const isShortage = fromB === "KHO RAU CỦ" && t.qtyReceived < t.qtyShipped;
    const isCorrective = fromB === "KHO RAU CỦ XỬ LÝ CHÊNH LỆCH CHUYỂN HÀNG";
    
    if (isShortage || isCorrective || t.matchedCorrectiveQty > 0 || status.statusText !== "Đủ") {
        console.log(`Date: ${t.date} | From: ${t.fromBranch} | To: ${t.toBranch}`);
        console.log(`   Trans: ${t.transferCode} | Code: ${t.itemCode} | Name: ${t.itemName}`);
        console.log(`   Excel Shipped: ${t.qtyShipped} | Excel Received: ${t.qtyReceived} | Matched corrective: ${t.matchedCorrectiveQty || 0}`);
        console.log(`   Calc: slChuyenKRC=${status.slChuyenKRC} | slNhanKRC=${status.slNhanKRC} | slBoSung=${status.slBoSung} | chenhLechConLai=${status.chenhLechConLai} | Status: ${status.statusText}`);
        console.log("--------------------------------------------------");
    }
});

File: inspect_app_yamazaki.py
import json
import re
import sys

sys.stdout.reconfigure(encoding='utf-8')

def norm(s):
    return str(s or "").strip().lower()

# 1. Load data.js
with open("data.js", "r", encoding="utf-8") as f:
    js_content = f.read()

# Extract window.initialTransfers
transfers_match = re.search(r"window\.initialTransfers\s*=\s*(\[.*?\])\s*;", js_content, re.DOTALL)
if not transfers_match:
    print("Error: Could not parse window.initialTransfers from data.js")
    exit(1)
    
transfers = json.loads(transfers_match.group(1))
print(f"Loaded {len(transfers)} transfers from data.js.")

# 2. Implement linkTransfers in Python
originals = []
correctives = []

for t in transfers:
    status = norm(t.get('docStatus', ''))
    if 'hủy' in status or 'huy' in status:
        continue
        
    fromB = norm(t.get('fromBranch', ''))
    if fromB == "kho rau củ":
        qtyRec = float(t.get('qtyReceived', 0))
        qtyShip = float(t.get('qtyShipped', 0))
        if qtyRec > qtyShip and qtyRec != -1:
            qtyRec = qtyShip
        originals.append({
            **t,
            'qtyReceived': qtyRec,
            'matchedCorrectiveQty': 0
        })
    elif fromB == "kho rau củ xử lý chênh lệch chuyển hàng":
        correctives.append({
            **t,
            'isMerged': False
        })

corrective_map = {}
for c in correctives:
    key = f"{norm(c.get('toBranch', ''))}_{norm(c.get('itemCode', ''))}"
    if key not in corrective_map:
        corrective_map[key] = []
    corrective_map[key].append(c)

from datetime import datetime

for orig in originals:
    key = f"{norm(orig.get('toBranch', ''))}_{norm(orig.get('itemCode', ''))}"
    candidates = corrective_map.get(key)
    if candidates:
        match = None
        # 1. Try to match by code link
        if orig.get('generatedDoc'):
            orig_gen = norm(orig.get('generatedDoc'))
            for c in candidates:
                if c.get('isMerged'):
                    continue
                c_code = norm(c.get('transferCode'))
                if c_code != "" and c_code in orig_gen:
                    match = c
                    break
        
        # 2. Fallback: match by branch, item, date proximity
        if not match and (float(orig.get('qtyShipped', 0)) > float(orig.get('qtyReceived', 0))):
            try:
                orig_date = datetime.strptime(orig.get('date'), '%Y-%m-%d')
                for c in candidates:
                    if c.get('isMerged'):
                        continue
                    try:
                        c_date = datetime.strptime(c.get('date'), '%Y-%m-%d')
                        diff_days = abs((c_date - orig_date).days)
                        if diff_days <= 3 and float(c.get('qtyShipped', 0)) > 0:
                            match = c
                            break
                    except Exception as ex:
                        pass
            except Exception as ex:
                pass
                
        if match:
            orig['matchedCorrectiveQty'] = float(match.get('qtyShipped', 0))
            match['isMerged'] = True

processed_transfers = originals + correctives

# 3. Implement calculateStatus in Python
def calculate_status(t):
    slChuyenKRC = float(t.get('qtyShipped', 0))
    slNhanKRC = float(t.get('qtyReceived', 0))
    
    fromB = norm(t.get('fromBranch', ''))
    if fromB == "kho rau củ" and slNhanKRC > slChuyenKRC and slNhanKRC != -1:
        slNhanKRC = slChuyenKRC
        
    chenhLech = 0
    slBoSung = 0
    chenhLechConLai = 0
    statusText = "Đủ"
    
    if slNhanKRC == -1:
        return {
            'slChuyenKRC': slChuyenKRC,
            'slNhanKRC': slNhanKRC,
            'chenhLech': 0,
            'slBoSung': 0,
            'chenhLechConLai': 0,
            'statusText': "Đang chuyển"
        }
        
    toB = norm(t.get('toBranch', ''))
    isCorrective = fromB == "kho rau củ xử lý chênh lệch chuyển hàng"
    isSupermarket = toB.startswith("kfm")
    
    chenhLech = slNhanKRC - slChuyenKRC
    
    if isCorrective:
        slBoSung = slChuyenKRC if isSupermarket else 0
        chenhLechConLai = chenhLech
    else:
        slBoSung = float(t.get('matchedCorrectiveQty', 0))
        if chenhLech < 0:
            chenhLechConLai = chenhLech + slBoSung
        elif chenhLech > 0:
            chenhLechConLai = chenhLech - slBoSung
        else:
            chenhLechConLai = 0
            
    if chenhLechConLai < 0:
        unit = norm(t.get('unit', ''))
        if unit == 'kg' and slChuyenKRC > 0 and (abs(chenhLechConLai) / slChuyenKRC) < 0.15:
            statusText = "Hao hụt"
        else:
            statusText = "Thiếu"
    elif chenhLechConLai > 0:
        statusText = "Dư"
    else:
        statusText = "Đủ"
        
    return {
        'slChuyenKRC': slChuyenKRC,
        'slNhanKRC': slNhanKRC,
        'chenhLech': chenhLech,
        'slBoSung': slBoSung,
        'chenhLechConLai': chenhLechConLai,
        'statusText': statusText
    }

# Filter and print Yamazaki items
codes = ['sp001222', 'sp001988']
print("\n--- YAMAZAKI TRANSFERS REPORT ---")
count = 0
for t in processed_transfers:
    if norm(t.get('itemCode')) in codes:
        status = calculate_status(t)
        fromB = norm(t.get('fromBranch'))
        isShortage = fromB == "kho rau củ" and float(t.get('qtyReceived', 0)) < float(t.get('qtyShipped', 0))
        isCorrective = fromB == "kho rau củ xử lý chênh lệch chuyển hàng"
        
        if isShortage or isCorrective or float(t.get('matchedCorrectiveQty', 0)) > 0 or status['statusText'] != "Đủ":
            count += 1
            print(f"Date: {t.get('date')} | From: {t.get('fromBranch')} | To: {t.get('toBranch')}")
            print(f"  Trans: {t.get('transferCode')} | Code: {t.get('itemCode')} | Name: {t.get('itemName')}")
            print(f"  Excel Shipped: {t.get('qtyShipped')} | Excel Received: {t.get('qtyReceived')} | Matched corrective: {t.get('matchedCorrectiveQty', 0)}")
            print(f"  Calc: slChuyen={status['slChuyenKRC']} | slNhan={status['slNhanKRC']} | slBoSung={status['slBoSung']} | chenhLechConLai={status['chenhLechConLai']} | Status: {status['statusText']}")
            print("-" * 50)
            
print(f"Printed {count} abnormal/corrective records.")
